### PR TITLE
[improve](serde) support complex type in write/read pb serde

### PR DESCRIPTION
**Pull request by zhangstar333**
## Proposed changes
support complex type and ip/jsonb in DataTypeSerDe::write_column_to_pb/read_column_from_pb function

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

File: gensrc/proto/types.proto [hunk 1]
```diff
@@ -188,6 +188,8 @@ message PValues {
     repeated string string_value = 11;
     repeated bytes bytes_value = 12;
     repeated PDateTime datetime_value = 13;
+    repeated PValues child_element = 14;
+    repeated int64 child_offset=15;
```
Comment (HappenLee) [inline gensrc/proto/types.proto:192]: = 15

File: be/src/vec/data_types/serde/data_type_decimal_serde.h [hunk 2]
```diff
@@ -154,10 +154,10 @@ template <typename T>
 Status DataTypeDecimalSerDe<T>::read_column_from_pb(IColumn& column, const PValues& arg) const {
     if constexpr (std::is_same_v<T, Decimal<Int128>> || std::is_same_v<T, Decimal128V3> ||
                   std::is_same_v<T, Decimal256> || std::is_same_v<T, Decimal<Int32>>) {
-        column.resize(arg.bytes_value_size());
+        column.reserve(arg.bytes_value_size());
```
Comment (HappenLee) [inline be/src/vec/data_types/serde/data_type_decimal_serde.h:157]: why change reserve and call emplace_back?

File: be/src/vec/data_types/serde/data_type_ipv4_serde.cpp [hunk 3]
```diff
@@ -87,6 +87,33 @@ Status DataTypeIPv4SerDe::deserialize_one_cell_from_json(IColumn& column, Slice&
...
 }
+Status DataTypeIPv4SerDe::write_column_to_pb(const IColumn& column, PValues& result, int start,
+                                             int end) const {
+    const auto& column_data = assert_cast<const ColumnIPv4&>(column);
+    result.mutable_string_value()->Reserve(end - start);
+    auto* ptype = result.mutable_type();
+    ptype->set_id(PGenericType::IPV4);
+    for (int i = start; i < end; ++i) {
+        IPv4Value ipv4_value(column_data.get_data()[i]);
+        result.add_string_value(ipv4_value.to_string());
```
Comment (HappenLee) [inline be/src/vec/data_types/serde/data_type_ipv4_serde.cpp:98]: why not use int but `to_string`

File: be/src/vec/data_types/serde/data_type_jsonb_serde.cpp [hunk 4]
```diff
@@ -239,6 +239,35 @@ Status DataTypeJsonbSerDe::read_one_cell_from_json(IColumn& column,
...
+    const auto& string_column = assert_cast<const ColumnString&>(column);
+    result.mutable_string_value()->Reserve(end - start);
+    auto* ptype = result.mutable_type();
+    ptype->set_id(PGenericType::JSONB);
+    for (size_t row_num = start; row_num < end; ++row_num) {
+        const auto& string_ref = string_column.get_data_at(row_num);
+        if (string_ref.size > 0) {
+            std::string json_string =
+                    JsonbToJson::jsonb_to_json_string(string_ref.data, string_ref.size);
+            result.add_string_value(json_string);
```
Comment (HappenLee) [inline be/src/vec/data_types/serde/data_type_jsonb_serde.cpp:253]: result.add_string_value(JsonbToJson::jsonb_to_json_string(string_ref.data, string_ref.size));

to call std::move

File: be/src/vec/data_types/serde/data_type_jsonb_serde.cpp [hunk 5]
```diff
@@ -239,6 +239,35 @@ Status DataTypeJsonbSerDe::read_one_cell_from_json(IColumn& column,
...
+        }
+    }
+    return Status::OK();
+}
 
+Status DataTypeJsonbSerDe::read_column_from_pb(IColumn& column, const PValues& arg) const {
+    auto& column_string = assert_cast<ColumnString&>(column);
+    column_string.reserve(arg.string_value_size());
+    for (int i = 0; i < arg.string_value_size(); ++i) {
+        JsonBinaryValue value;
```
Comment (HappenLee) [inline be/src/vec/data_types/serde/data_type_jsonb_serde.cpp:265]: better out the for loop

File: be/src/vec/data_types/serde/data_type_struct_serde.cpp [hunk 6]
```diff
@@ -413,5 +413,32 @@ Status DataTypeStructSerDe::write_column_to_orc(const std::string& timezone, con
...
+                                                              *child_elements[i], start, end));
+    }
+    return Status::OK();
+}
+
+Status DataTypeStructSerDe::read_column_from_pb(IColumn& column, const PValues& arg) const {
+    auto& struct_column = assert_cast<ColumnStruct&>(column);
+    DCHECK_EQ(struct_column.tuple_size(), arg.child_element_size());
+    for (size_t i = 0; i < struct_column.tuple_size(); ++i) {
+        RETURN_IF_ERROR(elemSerDeSPtrs[i]->read_column_from_pb(struct_column.get_column(i),
```
Comment (HappenLee) [inline be/src/vec/data_types/serde/data_type_struct_serde.cpp:437]: change to `elem_serde_ptrs`

**Comment by zhangstar333** [comment]
run buildall

File: be/src/vec/data_types/serde/data_type_number_serde.h [hunk 7]
```diff
@@ -109,47 +109,47 @@ template <typename T>
 Status DataTypeNumberSerDe<T>::read_column_from_pb(IColumn& column, const PValues& arg) const {
     if constexpr (std::is_same_v<T, UInt8> || std::is_same_v<T, UInt16> ||
                   std::is_same_v<T, UInt32>) {
-        column.resize(arg.uint32_value_size());
+        column.reserve(arg.uint32_value_size());
```
Comment (HappenLee) [inline be/src/vec/data_types/serde/data_type_number_serde.h:112]: column.reserve (column.size() + xxx);

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by BiteTheDDDDt — APPROVED** [review]

**Pull request merged**